### PR TITLE
Linkwarden

### PR DESCRIPTION
**Pull request by BrokenOnedroid**
### General information on project
This pull request proposes to add a new rock-on for the following project:
- name: Linkwarden
- website: https://linkwarden.app/
- description: Linkwarden is an open-source collaborative bookmark manager to collect, organize and preserve webpages.

### Information on docker image
- docker image: https://github.com/linkwarden/linkwarden/pkgs/container/linkwarden)
- is an official docker image available for this project?:  yes

### Checklist
- [x] Passes [JSONlint](https://jsonlint.com) validation
- [x] Entry added to `root.json` in _alphabetical_ order (for new rock-on only)
- [x] `"description"` object lists and links to the docker image used
- [x] `"description"` object provides information on the image's particularities (advantage over another existing rock-on for the same project, for instance)
- [x] `"website"` object links to project's main website

### CODE REVIEW

**Comment by Hooverdan96** [comment]
I have tested the installation:
image
image
image
image
image

After the install completed, I needed to create a new account, and subsequently was able to log in:
image

So, this works!

Since there are field-interdependencies w.r.t. password and port setting, I am suggesting some additional comments in the field descriptions. If you agree you should be able to just accept the suggestions directly from this thread.

File: linkwarden.json [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+                        "label": "PostgreSQL URL",
+                        "index": 3
+                    },
+                    "NEXTAUTH_SECRET": {
+                        "description": "A secure secret phrase.",
+                        "label": "NextAuth Secret",
+                        "index": 2
+                    },
+                    "NEXTAUTH_URL": {
+                        "description": "Base URL for Linkwarden. Set to: http://localhost:3000/api/v1/auth",
```
Comment (Hooverdan96) [inline linkwarden.json:37]:
```suggestion
                        "description": "Base URL for Linkwarden. Set to: http://localhost:<Web UI port defined earlier>/api/v1/auth",
```

File: linkwarden.json [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+                    "3000": {
+                        "description": "The port for accessing the Linkwarden Web UI.",
+                        "label": "Web UI Port",
+                        "host_default": 3000,
+                        "ui": true
+                    }
+                },
+                "volumes": {
+                    "/data/data": {
+                        "description": "Choose a Share for Linkwarden data storage.",
```
Comment (Hooverdan96) [inline linkwarden.json:21]:
```suggestion
                        "description": "Choose a Share for Linkwarden data storage. E.g., create a share named linkwarden-data.",
```

File: linkwarden.json [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+                "launch_order": 1,
+                "volumes": {
+                    "/var/lib/postgresql/data": {
+                        "description": "Choose a Share for PostgreSQL database storage.",
+                        "label": "PostgreSQL Storage"
+                    }
+                },
+                "environment": {
+                    "POSTGRES_PASSWORD": {
+                        "description": "Enter the password for the PostgreSQL 'postgres' super-user.",
```
Comment (Hooverdan96) [inline linkwarden.json:55]:
```suggestion
                        "description": "Enter the password for the PostgreSQL 'postgres' super-user that was entered previously as part of the 'PostgreSQL URL'",
```

**Review by Hooverdan96 — COMMENTED** [review]
I assume, we really don't have many options to prevent the duplicate entry/dependencies between the WebUI port and the BaseURL, nor for the Postgres superuser password and the connection string. Only for the Postgres connection I could see it to move to the `opts` object, unless you think there's a good use case for knowing the super user password for the database.

**Comment by Hooverdan96** [comment]
@BrokenOnedroid if you can review and commit the suggested changes, then I think this will be ready for final review and can be merged.

**Comment by BrokenOnedroid** [comment]
@Hooverdan96 done

**Review by Hooverdan96 — APPROVED** [review]
This looks good, the tooltips provide the right clarity I think. Another test installation seemed to have been successful (none of the previous configuration items changed, only text elements.

**Comment by Hooverdan96** [comment]
@phillxnet when you get a chance to merge and publish this. Thanks again @BrokenOnedroid for submitting this.

**Comment by phillxnet** [comment]
@BrokenOnedroid & @Hooverdan96 Thanks for preparing this Rock-on. It looks like a nice addition.

It does rather highlight some of our current limitations however: with the install being super fragile re by-hand copying from tooltip suggestions. But this can all be improved upon as we up our Rock-on facilities over time. And with some pending examples of a preferred presentation that we hope to form into a guide regarding label naming and a new format to be expected across all Rock-on labels.

The `DATABASE_URL` was particularly challenging re input here. Especially with its associated password appearing later in the list (again a limitation on our end). It was also not clear what NEXTAUTH_* was all about. I.e. we have not user level explanation here and that is what Rock-ons are about in essence. I.e. "A secure secret phrase." for what?

I.e. on the label front:
- NEXTAUTH_SECRET: "NextAuth Secret"
- NEXTAUTH_URL: "NextAuth URL"

I know one can look these up - but in part our place as a wrapper is not to have the env vars presented bare with nothing but an upstream reference to look them up. We try to explain what each is for. If it's required to surface the env var we are setting then so be it: i.e. it's technical and so one needs to inform the user what element in upstream it references. But we do present a summary (pre-install) of how we translate what is input to the docker level.

Again function trumps form, and this looks to work. Plus it's far easier to refine something that exists than to have to start from scratch. So we are at least started here and I'll merge and publish accordingly. But more Rock-ons with self inconsistent labels/suggestions etc. And only a case shift of an environmental variable as presented at the docker command level rather makes this more tricky than a copy-past of a docker command. Ideally we need to have a little more refinement.

But all in, we have a start. But it represents, not like a goodly number of our other Rock-ons mind :), a potential technical dept. I.e. there is work to be done on the user-text side. But we also have work to do in communicating our preferred format for example labels. And as always this is ongoing: we we are getting there.

For all involved: do not take this overly negatively, we are just moving to up our presentation game also here. And for that we need a doc guide on what is to be expected for labels etc.

However - nice to have this Rock-on submitted and in place (about to merge and publish). And thanks again for all efforts getting it this far.

**Comment by phillxnet** [comment]
@Hooverdan96 We could likely use some kind of random password/passphrase place-holder: which when encountered in a Rock-on definition is substituted for an actual random pass. And can be re-used multiple times in a Rock-on definition. Probably needs a new named element with some definition on the type of password/passphrase to generate. Tricky to be backward compatible here thought. But we could introduce some minimum Rockstor version element: surfaced to the user. Maybe something for the next or next but one testing channel phase.

**Comment by phillxnet** [comment]
PRODUCTION published

**Comment by Hooverdan96** [comment]
@phillxnet agreed on your points. That has always been a challenge, that one cannot really copy from the tooltips. I think it would make sense, at the least, to introduce defaults for all elements that can be surfaced (I think we have an issue open related to this). That way the copy/paste could be avoided or at least minimized. Even if the default shows (like in this one) some element that has to be changed for sure, because of dependency.

Secondly, based on your thoughts above, maybe introduce another json element, as optional, that can be handled like a variable, i.e. it's exposed once, the user defines it, and then can be mapped into other containers and their variables. Making that optional would possibly allow the backwards compatibility and give time to eventually convert all Rockons over time to take advantage of these things where applicable.

**Comment by Hooverdan96** [comment]
With regards to default tags, the rockstor-core repot, we currently have:
[#2094](https://github.com/rockstor/rockstor-core/issues/2094)
[#2012](https://github.com/rockstor/rockstor-core/issues/2012)

**Comment by phillxnet** [comment]
@Hooverdan96 Re:

> ... maybe introduce another json element, as optional, that can be handled like a variable, i.e. it's exposed once, the user defines it, and then can be mapped into other containers and their variables.

Well put. Yes; that would be super handy, and along with a default=autopassword, or the like and it can pre-set an internal password between containers such as is more common re all these postgres containers etc. We could also have it internal only (optionally) so there is no interaction/input required if need be.
